### PR TITLE
Implement Text Selection handles in TextBox for touch input

### PR DESCRIPTION
**Pull request by emmauss**
## What does the pull request do?
Implements text selection handles in textbox that allow changing text selection by dragging. This is activated by non-mouse input, i.e. touch and pen. This provides similar text selection behavior to what users are familiar with on android and iOS.

https://github.com/AvaloniaUI/Avalonia/assets/5307160/53da0119-eb36-4294-975f-59b9fef7e308

In addition, the textbox context menu has been simplified and optimized for mobile.

https://github.com/AvaloniaUI/Avalonia/assets/5307160/21f44470-e8dc-44de-99a0-0c2922b80c2c

Though what I've planned for this pr is all in, it's still subject to change, so comments on implementation details and recommendations are welcome.

## What is the current behavior?
Currently, text can only be be selected by dragging the caret along the text. This makes it very inaccurate on touch devices, and text selection is easily reset by accidental touches.

## How was the solution implemented (if it's not obvious)?
SelectionHandles must be drawn over all other controls, and though visibility of the handles is affected by the current text cursor, it shouldn't be clipped. As such, a new visual layer with a higher z-index is used to host the handles. This has the side effect of selection handles drawing over hosted popups, which are drawn in the OverlayLayer. It's possible to add text controls to the OverlayLayer, as such, that layer is not suitable to host selection handles.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
This is a behavioral breaking change.

### CODE REVIEW

**Comment by robloo** [comment]
Shouldn't this be done with adorners? They are specifically designed for cases like this.

I'm a little uncomfortable with how this was implemented with new canvas controls overlaying everything else. It doesn't seem to match XAML ideas.

The graphical design looks good though and this feature is a big plus for mobile!

**Comment by emmauss** [comment]
> Shouldn't this be done with adorners? They are specifically designed for cases like this.
> 
> I'm a little uncomfortable with how this was implemented with new canvas controls overlaying everything else. It doesn't seem to match XAML ideas.
> 
> The graphical design looks good though and this feature is a big plus for mobile!

Adorners currently can't be used in the any layer above the AdronerLayer, which includes OverlayLayer. It also has a composition bug, https://github.com/AvaloniaUI/Avalonia/issues/10409 .

**Comment by robloo** [comment]
A few more comments:

 1.  > It's possible to add text controls to the OverlayLayer, as such, that layer is not suitable to host selection handles.

      I believe that's too severe of a limitation. Several input-type controls have a "popup" that will appear as an overlay on certain platforms. This would not allow modification of text in this scenario using the seleciton handles which seems to be a rather big architecture flaw. This indicates another approach is needed.

 2. The Cut/Copy/Paste menu is going to need to be customized for mobile platforms I think. Instead of a vertical layout of the commands iOS and Android both use a horizontal layout. This is important due to mobile form factor and screen sizes. I think this is fairly easy to do with the new OnPlatform markup extension.

 3. Looking at some iOS documentation it appears they also side-step this issue by placing the overlay cut/copy/paste buttons ABOVE the selection in most cases. Again using a Y-position offset to avoid any intersection.

**Comment by emmauss** [comment]
> > Adorners currently can't be used in the any layer above the AdornerLayer, which includes OverlayLayer
> 
> Why would the selection handles need to appear over the overlay layer? In Android they do the following:
> 
>     1. Selection handles are BELOW the overlay layer which contains cut/copy/paste (this can be confirmed by moving the overlay manually). So they accept adorner below overlay layer in concept. This isn't a problem because:
> 
>     2. When changing the position of the selection handles, the overlay cut/copy/paste controls hide themselves. They will re-appear after a few millisecond delay again when the selection handles are no longer being moved.
> 
>     3. The overlay cut/copy/paste controls are shown in a Y-position below the selection handles. The position of the overlay needs to account for the selection handles (this alone solves most of the issues)
> 
> 
> Someone can check iOS as well but simply copying Android would allow Adorner usage.
> 
> > It also has a composition bug, #10409 .
> 
> There a few other issues for adorners as well: #10701 and #11232. However, those bugs should not drive a "hacky" solution for a core feature. The correct path is to fix adorners. We don't want a temporary solution that will stick around in the code base for years I think?

I know of 1 popular avalonia control toolkit that is the overlay layler for managed dialogs, ie content dialogs. These dialogs can host text boxes. And the overlay layer is above the adorner layer.

**Comment by emmauss** [comment]
@robloo Android uses popups for their copy-paste menu. These are windows on their own, similar to desktop

**Comment by robloo** [comment]
> Android uses popups for their copy-paste menu. These are windows on their own, similar to desktop

Ok, even if they are 'windows' rather than an overlay my point stands about z-index. The selection handles are underneath them as shown below. If it works for Android I have to think it would work for Avalonia.... however

image

> I know of 1 popular avalonia control toolkit that uses the overlay layler for managed dialogs, ie content dialogs. These dialogs can host text boxes. And the overlay layer is above the adorner layer.

Yea, true, the adorner layer would still have to render above the overlay layer so the selection handles could appear in all cases like this. I was hoping this could be avoided but it can't - overlay hosted controls need selection handles too in some cases.

This also might be a little bit new territory since UWP didn't have adorners. Perhaps that's why they don't fully work in composition renderer now.

Perhaps the solution is to:
 1. Make adorners work again
 2. Render adorners above overlay layer. Actually, any adorner created in an overlay should just work so there is clearly a missing feature here.
 3. Make cut/copy/paste buttons horizontal on mobile to save Y-axis space
 4. Move cut/copy/paste buttons by:
     a. Hiding them when selection handles are being dragged
     b. Only show cut/copy/paste overlay above/below the selection box in the Y-axis

TL;DR: I still think adorners should be used. They were designed exactly for things like this.

**Comment by emmauss** [comment]
> > Android uses popups for their copy-paste menu. These are windows on their own, similar to desktop
> 
> Ok, even if they are 'windows' rather than an overlay my point stands about z-index. The selection handles are underneath them as shown below. If it works for Android I have to think it would work for Avalonia.... however
> 
> image
> 
> > I know of 1 popular avalonia control toolkit that uses the overlay layler for managed dialogs, ie content dialogs. These dialogs can host text boxes. And the overlay layer is above the adorner layer.
> 
> Yea, true, the adorner layer would still have to render above the overlay layer so the selection handles could appear in all cases like this. I was hoping this could be avoided but it can't - overlay hosted controls need selection handles too in some cases.
> 
> This also might be a little bit new territory since UWP didn't have adorners. Perhaps that's why they don't fully work in composition renderer now.
> 
> Perhaps the solution is to:
> 
>     1. Make adorners work again
> 
>     2. Render adorners above overlay layer. Actually, any adorner created in an overlay should just work so there is clearly a missing feature here.
> 
>     3. Make cut/copy/paste buttons horizontal on mobile to save Y-axis space
> 
>     4. Move cut/copy/paste buttons by:
>        a. Hiding them when selection handles are being dragged
>        b. Only show cut/copy/paste overlay above/below the selection box in the Y-axis
> 
> 
> TL;DR: I still think adorners should be used. They were designed exactly for things like this.

3 and 4 are on my list of things to do for touch. 1 and 2 will need @kekekeks input. I think OverlayLayer was added after Adorners, so it was placed above so as not to break stuff.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040570-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by emmauss** [comment]
@robloo Added mobile themed context flyouts for textbox

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040578-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Controls/Primitives/TextSelectionHandle.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using Avalonia.Input;
+using Avalonia.Interactivity;
+using Avalonia.Media;
+
+namespace Avalonia.Controls.Primitives
+{
+    public class TextSelectionHandle : TemplatedControl
```
Comment (MrJul) [inline src/Avalonia.Controls/Primitives/TextSelectionHandle.cs:8]: Can this class inherit from `Thumb`? Both types seem have a lot of similar code and behavior.

File: src/Avalonia.Controls/Primitives/TextSelectionHandle.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            remove { RemoveHandler(DragDeltaEvent, value); }
+        }
+
+        internal event EventHandler<VectorEventArgs> DragCompleted
+        {
+            add { AddHandler(DragCompletedEvent, value); }
+            remove { RemoveHandler(DragCompletedEvent, value); }
+        }
+
+        internal SelectionHandleType SelectionHandleType
```
Comment (MrJul) [inline src/Avalonia.Controls/Primitives/TextSelectionHandle.cs:39]: The type `SelectionHandleType` is public, but this property isn't. Shouldn't it? Otherwise there's no real point of it being a `StyledProperty`.

File: src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+        public void MoveHandlesToSelection()
+        {
+            if (_presenter == null || _textBox == null || _startHandle.IsDragging || _endHandle.IsDragging)
+            {
+                return;
+            }
+
+            var hasSelection = _textBox.SelectionStart != _textBox.SelectionEnd;
+            if (_caretHandle != null)
```
Comment (MrJul) [inline src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs:195]: The three handles are never `null` as they're all assigned only once in constructor (they can even be made readonly).

File: src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+            if (e.Pointer.Type != PointerType.Mouse)
+            {
+                ShowHandles = true;
+
+                MoveHandlesToSelection();
+                EnsureVisible();
+            }
+        }
+
+        private void _textBox_LayoutUpdated(object? sender, EventArgs e)
```
Comment (MrJul) [inline src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs:326]: Nit: naming `TextBoxLayoutUpdated` and `TextBoxPropertyChanged` to match the other handlers.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 5]
```diff
@@ -286,6 +291,8 @@ public int SelectionEnd
 
         protected override bool BypassFlowDirectionPolicies => true;
 
+        internal TextSelectionCanvas? Canvas { get => _canvas; set => _canvas = value; }
```
Comment (MrJul) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs:294]: Nit: the field can be removed and an auto property used instead

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 6]
```diff
@@ -818,11 +839,23 @@ protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
...
             _caretTimer.Tick += CaretTimerTick;
+
+            if (_canvas == null)
+            {
+                _canvas = new TextSelectionCanvas();
+            }
+
+            _layer = TextSelectorLayer.GetTestLayer(this);
+            _layer?.Add(_canvas);
+            _canvas?.SetPresenter(this);
```
Comment (MrJul) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs:850]: `_canvas` can't be null here, it's be assigned above

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 7]
```diff
@@ -818,11 +839,23 @@ protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
...
         }
 
         protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
         {
             base.OnDetachedFromVisualTree(e);
+            if (_canvas is { } c)
+                _layer?.Remove(c);
+            _canvas?.SetPresenter(null);
```
Comment (MrJul) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs:858]: Nit: can be move inside the above if since it's already making the null check, e.g. `c.SetPresenter(null)`

File: src/Avalonia.Controls/TextBox.cs [hunk 8]
```diff
@@ -1470,7 +1459,7 @@ protected override void OnPointerPressed(PointerPressedEventArgs e)
             var text = Text;
             var clickInfo = e.GetCurrentPoint(this);
 
-            if (text != null && clickInfo.Properties.IsLeftButtonPressed &&
+            if (text != null && (e.Pointer.Type == PointerType.Mouse || e.ClickCount >= 2) && clickInfo.Properties.IsLeftButtonPressed &&
```
Comment (MrJul) [inline src/Avalonia.Controls/TextBox.cs:1462]: I'm not sure what the new condition does, but it seems incompatible with the code inside the `if` branch: `isTouch && e.ClickCount == 1` below can now never be `true`

File: src/Avalonia.Controls/TextBox.cs [hunk 9]
```diff
@@ -1560,7 +1550,7 @@ protected override void OnPointerMoved(PointerEventArgs e)
             }
 
             // selection should not change during pointer move if the user right clicks
-            if (e.Pointer.Captured == _presenter && e.GetCurrentPoint(this).Properties.IsLeftButtonPressed)
+            if (e.Pointer.Captured == _presenter && e.Pointer.Type == PointerType.Mouse && e.GetCurrentPoint(this).Properties.IsLeftButtonPressed)
```
Comment (MrJul) [inline src/Avalonia.Controls/TextBox.cs:1553]: Same as above, there's touch specific code inside this method that's can never be called anymore

File: src/Avalonia.Controls/TextBox.cs [hunk 10]
```diff
@@ -1684,7 +1714,7 @@ protected override void OnPointerReleased(PointerReleasedEventArgs e)
 
                 if(SelectionStart != SelectionEnd)
                 {
-                    RaiseEvent(new ContextRequestedEventArgs(e));
+                    _presenter.Canvas?.ShowContextMenu();
```
Comment (MrJul) [inline src/Avalonia.Controls/TextBox.cs:1717]: Doesn't this change prevent the user from customizing/handling the context menu manually?

File: src/Avalonia.Themes.Fluent/Controls/TextBox.xaml [hunk 11]
```diff
@@ -21,11 +21,23 @@
...
     <MenuItem x:Name="TextBoxContextFlyoutCopyItem" Header="Copy" Command="{Binding $parent[TextBox].Copy}" IsEnabled="{Binding $parent[TextBox].CanCopy}" InputGesture="{x:Static TextBox.CopyGesture}"/>
     <MenuItem x:Name="TextBoxContextFlyoutPasteItem" Header="Paste" Command="{Binding $parent[TextBox].Paste}" IsEnabled="{Binding $parent[TextBox].CanPaste}" InputGesture="{x:Static TextBox.PasteGesture}"/>
   </MenuFlyout>
+  <MenuFlyout x:Key="MobileTextBoxContextFlyout" FlyoutPresenterTheme="{StaticResource HorizontalFlyout}">
+    <MenuItem x:Name="MobileTextBoxContextFlyoutCutItem" Classes="textMenu" Header="Cut" Command="{Binding $parent[TextBox].Cut}" IsEnabled="{Binding $parent[TextBox].CanCut}" />
+    <MenuItem x:Name="MobileTextBoxContextFlyoutCopyItem" Classes="textMenu" Header="Copy" Command="{Binding $parent[TextBox].Copy}" IsEnabled="{Binding $parent[TextBox].CanCopy}" />
+    <MenuItem x:Name="MobileTextBoxContextFlyoutPasteItem" Classes="textMenu" Header="Paste" Command="{Binding $parent[TextBox].Paste}" IsEnabled="{Binding $parent[TextBox].CanPaste}" />
```
Comment (MrJul) [inline src/Avalonia.Themes.Fluent/Controls/TextBox.xaml:39]: Classes are usually `kebab-case`, e.g. `text-menu`

**Comment by MrJul** [comment]
> There a few other issues for adorners as well: #10701 and #11232. However, those bugs should not drive a "hacky" solution for a core feature. The correct path is to fix adorners. We don't want a temporary solution that will stick around in the code base for years I think?

I completely agree with this statement. We're adding a public layer here to work around adorners issues, but they're really made for this case. Even when adorners get fixed, the current solution might have to stay, which really isn't ideal.

----

Apart from that, this PR is a great and much needed improvement for mobile!

File: src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+                if (_textBox != null)
+                {
+                    _textBox.AddHandler(TextBox.TextChangingEvent, TextChanged, handledEventsToo: true);
+                    _textBox.AddHandler(KeyDownEvent, TextBoxKeyDown, handledEventsToo: true);
+                    _textBox.AddHandler(LostFocusEvent, TextBoxLostFocus, handledEventsToo: true);
+                    _textBox.AddHandler(PointerReleasedEvent, TextBoxPointerReleased, handledEventsToo: true);
+                    _textBox.AddHandler(Gestures.HoldingEvent, TextBoxHolding, handledEventsToo: true);
+
+                    _textBox.PropertyChanged += _textBox_PropertyChanged;
+                    _textBox.LayoutUpdated += _textBox_LayoutUpdated;
```
Comment (MrJul) [inline src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs:247]: I know this was discussed during development, but `LayoutUpdated` is raised directly by the layout manager itself: even if some unrelated control in an unrelated part of the view has been re-arranged, the event will trigger for all textboxes, potentially causing performances problems.

It's needed here to know when the text box moves (e.g. it's inside a scroll viewer), to move the handles along with it.

I became aware two days ago of the `EffectiveViewportChanged` event. It should solve the problem nicely: it's only raised on the control itself, when its visible viewport has changed. It should be perfect for the use case here.

Comment (emmauss) [reply src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs]: EffectiveViewportChanged works. thanks.

File: src/Avalonia.Controls/TextBox.cs (hunk 10 above)
Comment (emmauss) [reply src/Avalonia.Controls/TextBox.cs]: It now internally calls `ContextRequestedEventArgs`.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040840-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by emmauss** [comment]
Cleaned up and ready for review. There's a version of this branch which uses adorners here, https://github.com/AvaloniaUI/Avalonia/tree/text_selector_adorner .

File: src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml [hunk 13]
```diff
@@ -32,4 +32,39 @@
       </ControlTemplate>
     </Setter>
   </ControlTheme>
+  <ControlTheme x:Key="MobileHorizontalMenuFlyoutPresenter" TargetType="MenuFlyoutPresenter">
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml:35]: I don't think it's necessary to implement my idea in this PR, but in general, we can have TextBox.SelectionFlyout property, keeping it consistent with other Flyout props and UWP.
https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.selectionflyout?view=winrt-22621
And TextCommandBarFlyout class that has a default selection 
https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textcommandbarflyout?view=winrt-22621

File: src/Avalonia.Themes.Fluent/Controls/TextBox.xaml [hunk 14]
```diff
@@ -102,7 +114,7 @@
     <Setter Property="Padding" Value="{DynamicResource TextControlThemePadding}" />
     <Setter Property="FocusAdorner" Value="{x:Null}" />
     <Setter Property="ScrollViewer.IsScrollChainingEnabled" Value="True" />
-    <Setter Property="ContextFlyout" Value="{StaticResource DefaultTextBoxContextFlyout}" />
+    <Setter Property="ContextFlyout" Value="{OnFormFactor Desktop={StaticResource DefaultTextBoxContextFlyout}, Mobile={StaticResource MobileTextBoxContextFlyout}}" />
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/TextBox.xaml:117]: If anything, it should differentiate on input type, but can be changed after media queries merged.

File: src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml (hunk 13 above)
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml:35]: In general, UWP has ContextFlyout and SelectionFlyout independent:
image

image

Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml:35]: For now, let's keep it simpler. It doesn't seem to be a big breaking change later.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 15]
```diff
@@ -818,11 +837,25 @@ protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
             base.OnAttachedToVisualTree(e);
 
             _caretTimer.Tick += CaretTimerTick;
+
+            if (TextSelectionHandleCanvas == null)
+            {
+                TextSelectionHandleCanvas = new TextSelectionHandleCanvas();
+            }
+
+            _layer = TextSelectorLayer.GetTestLayer(this);
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs:846]: Test layer?

File: src/Avalonia.Controls/Primitives/SelectionHandleType.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Avalonia.Controls.Primitives
+{
+    public enum SelectionHandleType
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Primitives/SelectionHandleType.cs:3]: nit: docs

File: src/Avalonia.Controls/Primitives/TextSelectionHandle.cs [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+{
+    public class TextSelectionHandle : Thumb
+    {
+        internal SelectionHandleType SelectionHandleType { get; set; }
+
+        private Point _startPosition;
+
+        internal Point IndicatorPosition => IsDragging ? _startPosition.WithX(_startPosition.X) + _delta : Bounds.Position.WithX(Bounds.Position.X).WithY(Bounds.Y);
+
+        public bool IsDragging { get; private set; }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Primitives/TextSelectionHandle.cs:16]: It probably should be a direct property or private/internal. As otherwise developers would want to use it in styling.

File: src/Avalonia.Controls/Primitives/SelectionHandleType.cs (hunk 16 above)
Comment (maxkatz6) [reply src/Avalonia.Controls/Primitives/SelectionHandleType.cs:3]: You also can run a formatting tool on new files.

File: src/Avalonia.Themes.Simple/Controls/TextSelectionHandle.xaml [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+                </GeometryGroup>
+              </PathIcon.Data>
+            </PathIcon>
+          </Grid>
+        </ControlTemplate>
+      </Setter.Value>
+    </Setter>
+    <Style Selector="^.caret /template/ PathIcon#PART_HandlePathIcon">
+      <Setter Property="Data">
+        <GeometryGroup FillRule="NonZero">
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Simple/Controls/TextSelectionHandle.xaml:36]: I think GeometryGroup can be reused and saved in resources. But not sure.

Comment (maxkatz6) [reply src/Avalonia.Themes.Simple/Controls/TextSelectionHandle.xaml:36]: Here and above

File: src/Avalonia.Themes.Fluent/Controls/TextBox.xaml [hunk 19]
```diff
@@ -21,11 +21,23 @@
...
 
-  <MenuFlyout x:Key="DefaultTextBoxContextFlyout" Placement="Bottom">
+  <ControlTheme x:Key="HorizontalFlyout" TargetType="MenuFlyoutPresenter" BasedOn="{StaticResource {x:Type MenuFlyoutPresenter}}">
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/TextBox.xaml:24]: Is it used anywhere?

File: src/Avalonia.Themes.Fluent/Controls/MenuItem.xaml [hunk 20]
```diff
@@ -199,6 +198,113 @@
...
 
+    <Style Selector="^:separator">
+      <Setter Property="Template">
+        <ControlTemplate>
+          <Separator/>
+        </ControlTemplate>
+      </Setter>
+    </Style>
+  </ControlTheme>
+  <ControlTheme x:Key="MobileTextMenuItem" TargetType="MenuItem">
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/MenuItem.xaml:209]: How much different are these themes to create a completely new one? I think we could reuse the template with BasedOn, and add small style inside to adjust its look.

Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/MenuItem.xaml:209]: We also have another MenuItem style for Menu (horizontal bar). While I don't expect it to be the same, we definitely can reuse some base styles.

File: src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml [hunk 21]
```diff
@@ -32,4 +32,39 @@
...
+    <Setter Property="Padding" Value="{DynamicResource MenuFlyoutPresenterThemePadding}" />
+    <Setter Property="MinWidth" Value="{DynamicResource MobileMenuFlyoutThemeMinWidth}" />
+    <Setter Property="ScrollViewer.HorizontalScrollBarVisibility" Value="Auto" />
+    <Setter Property="ScrollViewer.VerticalScrollBarVisibility" Value="Auto" />
+    <Setter Property="CornerRadius" Value="{DynamicResource OverlayCornerRadius}" />
+    <Setter Property="ItemsPanel">
+      <ItemsPanelTemplate>
+        <StackPanel Orientation="Horizontal"/>
+      </ItemsPanelTemplate>
+    </Setter>
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml:48]: Isn't this setter the most important different from a vertical presenter? It also can be reused with BasedOn I believe.

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml [hunk 22]
```diff
@@ -207,6 +207,11 @@
       <StaticResource x:Key="MenuFlyoutPresenterBackground" ResourceKey="SystemControlTransientBackgroundBrush" />
       <StaticResource x:Key="MenuFlyoutPresenterBorderBrush" ResourceKey="SystemControlTransientBorderBrush" />
       <Thickness x:Key="MenuFlyoutPresenterBorderThemeThickness">1</Thickness>
+      <!-- Mobile Resources for MenuFlyout.xaml -->
+      <x:Double x:Key="MobileMenuFlyoutThemeMinWidth">32</x:Double>
+      <x:Double x:Key="MobileMenuFlyoutSeparatorThemeWidth">1</x:Double>
+      <Thickness x:Key="MobileMenuFlyoutItemMargin">2,0,2,0</Thickness>
+      <Thickness x:Key="MobileMenuFlyoutItemThemePaddingNarrow">8,4,8,7</Thickness>
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml:214]: Maybe instead of "mobile" we should use "horizontal" naming for that.
Text selection handles and text selection flyouts aren't mobile only. It's essentially touch-specific, and there are a lot of touch devices on Windows/embed linux.

**Comment by maxkatz6** [comment]
I tested it on my iPad, and text editing really feels way better than before! Good job.

File: src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml (hunk 13 above)
Comment (emmauss) [reply src/Avalonia.Themes.Fluent/Controls/MenuFlyoutPresenter.xaml:35]: This sounds like a great idea. There may be some issues with how to trigger both separately, as the concept of different menus for different intents on edit fields doesn't exist on mobile. Maybe Hold to trigger the default menu, and tapping the handles or ending a selection gesture to trigger the selection menu.

File: src/Avalonia.Themes.Fluent/Controls/TextBox.xaml (hunk 14 above)
Comment (emmauss) [reply src/Avalonia.Themes.Fluent/Controls/TextBox.xaml:117]: This could be solved by the earlier  suggestion for a SelectionFlyout. I don't think media queries can differentiate by input.

File: src/Avalonia.Controls/Primitives/TextSelectionHandle.cs (hunk 17 above)
Comment (emmauss) [reply src/Avalonia.Controls/Primitives/TextSelectionHandle.cs:16]: Made internal.

**Comment by emmauss** [comment]
@maxkatz6 Addressed changes requested.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs (hunk 15 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: `TextSelectionLayer.GetLayer()`

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml [hunk 23]
```diff
@@ -991,6 +996,12 @@
       <StaticResource x:Key="MenuFlyoutPresenterBackground" ResourceKey="SystemControlTransientBackgroundBrush" />
       <StaticResource x:Key="MenuFlyoutPresenterBorderBrush" ResourceKey="SystemControlTransientBorderBrush" />
       <Thickness x:Key="MenuFlyoutPresenterBorderThemeThickness">1</Thickness>
+      <Thickness x:Key="MobileMenuFlyoutItemChevronMargin">0,0,0,0</Thickness>
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml:999]: missed one

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml [hunk 24]
```diff
@@ -991,6 +996,12 @@
...
+      <!-- Mobile Resources for MenuFlyout.xaml -->
+      <x:Double x:Key="HorizontalMenuFlyoutThemeMinWidth">32</x:Double>
+      <x:Double x:Key="HorizontalMenuFlyoutSeparatorThemeWidth">1</x:Double>
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml:1002]: I think it can be NaN instead of 1

**Review by maxkatz6 — APPROVED** [review]
LGTM

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml (hunk 23 above)
Comment (emmauss) [reply src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml]: Removed, as it's unused.

**Comment by MrJul** [comment]
I've just retested on Android (physical): the caret itself isn't draggable anymore when the handle(s) are visible.
Without selection, this prevents from moving the caret around with the finger (without using the handle).
With selection, it prevents from tapping on text to unselect / set the caret elsewhere.

File: src/Avalonia.Controls/TextBox.cs (hunk 8 above)
Comment (MrJul) [reply src/Avalonia.Controls/TextBox.cs:1462]: Unresolving this as dragging the caret itself doesn't work (see my comment below).

**Comment by emmauss** [comment]
> I've just retested on Android (physical): the caret itself isn't draggable anymore when the handle(s) are visible. Without selection, this prevents from moving the caret around with the finger (without using the handle). With selection, it prevents from tapping on text to unselect / set the caret elsewhere.

Weird. Could you record with the newest commit and post it?

**Comment by MrJul** [comment]
https://github.com/AvaloniaUI/Avalonia/assets/1623034/630ac615-85d3-443e-84e2-0cf8d923a9b0

Here it is. Note that the caret still moves, but only when I'm releasing my finger. In all other apps, the caret moves with the finger (and displays a zoomed text, but that's a different improvement).

**Comment by MrJul** [comment]
To clarify the "with selection" part: once you have a selection, you *can* unselect/move the caret but it takes two taps: one to dismiss the context menu (that opens automatically) and another one to move the caret elsewhere. One tap should be sufficient to both close the menu and move the caret (at least that's the behavior in all other apps).

That said, I understand if you want that change to be part of another PR. The current behavior is already way, way better than before.

**Comment by emmauss** [comment]
> 2023-10-19-15-42-07.mp4
> 
> Here it is. Note that the caret still moves, but only when I'm releasing my finger. In all other apps, the caret moves with the finger (and displays a zoomed text, but that's a different improvement).

Ok, I didn't account for that. There may be issues with scrolling, but I'll try to alleviate it.

**Comment by emmauss** [comment]
> To clarify the "with selection" part: once you have a selection, you _can_ unselect/move the caret but it takes two taps: one to dismiss the context menu (that opens automatically) and another one to move the caret elsewhere. One tap should be sufficient to both close the menu and move the caret (at least that's the behavior in all other apps).
> 
> That said, I understand if you want that change to be part of another PR. The current behavior is already way, way better than before.

This behavior would need https://github.com/AvaloniaUI/Avalonia/pull/13351 first, so will update after that's merged.

**Comment by emmauss** [comment]
@MrJul Should be good now. could you try again?

**Comment by MrJul** [comment]
> @MrJul Should be good now. could you try again?

Just tried for a few minutes, touch text selection and caret moving feels quite good, kudos!

You even fixed the outer scrollviewer preventing proper touch scrolling inside textboxes after a while (#13351 I presume), which was quite annoying.

**Comment by robloo** [comment]
Awesome work again!

Finally, what was decided about adorners? Will this be adjusted later when adorners work in overlays and the bugs are fixed?

**Comment by hez2010** [comment]
Several issues after testing on a real Android device:
- It's very hard to enter the selection mode, in most cases the selection will be dismissed immediately.
- On a multiple line textbox with a scrollbar, you cannot scroll the text area to adjust selection carets and the selection will dismiss after you scroll the textarea.

**Comment by emmauss** [comment]
> Several issues after testing on a real Android device:
> 
>     * It's very hard to enter the selection mode, in most cases the selection will be dismissed immediately.
> 
>     * On a multiple line textbox with a scrollbar, you cannot scroll the text area to adjust selection carets and the selection will dismiss after you scroll the textarea.

Selection mode is entered by either tapping the text to set the handle, then tapping the handle, or holding on the text. Drag selection for touch has been removed due to how it interferes with gestures.
Scrolling on a multiline textbox is done simply by moving the touch vertically when you press. Moving it horizontally immediately you press will disable scrolling for that pointer and instead allow changing the caret position. For a single line textbox that scrolls horizontally, using the handles to move selection is sufficient.
